bundler/lib/bundler/ruby_version.rb

```diff
@@ -23,7 +23,7 @@ def initialize(versions, patchlevel, engine, engine_version)
       #   specified must match the version.
 
       @versions = Array(versions).map do |v|
-        op, v = Gem::Requirement.parse(v)
+        op, v = Gem::Requirement.parse(v.tr("-", "."))
```

Comment:
Would it be possible to move this into `PATTERN` instead? Whole parsing logic is currently in there.

Reply:
Not really sure if modifying `PATTERN` will help. `3.3.0-preview2` does not violate any regex matching for `PATTERN`. Preview version format is `3.3.0-preview2` while `rubygems` repository works correctly with `3.3.0.preview2`. I think `3.3.0-preview2` does not work beacuse of below line in [version.rb](https://github.com/rubygems/rubygems/blob/master/lib/rubygems/version.rb#L229).

````ruby
@version = version.to_s.strip.gsub("-",".pre.")
```` 

This line converts `3.3.0-preview2` to `3.3.0.pre.preview2`, which is an incorrect Ruby version and causes below error while executing `bundle install`.

````ruby
Your Ruby version is 3.3.0.preview2, but your Gemfile specified 3.3.0.pre.preview2
````

Will dig a little deeper to see how this can be fixed using `PATTERN`.

Reply:
If PATTERN is not the best place to do this, just please wrap this into own method with documentation. This way it is just replacing `-` with `.` inline in constructor with no explanation.

Reply:
:thinking: changing ruby version part of PATTERN `(\d+\.\d+\.\d+(?:\.\S+)?)` into `(\d+\.\d+\.\d+(?:\.?-?\S+)?)` seems properly parsing the `3.3.0-preview2` versions. No idea if that breaks other parts.

Reply:
> I think 3.3.0-preview2 does not work beacuse of below line in [version.rb](https://github.com/rubygems/rubygems/blob/master/lib/rubygems/version.rb#L229).

I'll check with this as well, thanks for the info.

Reply:
I created new branch and just changed the PATTERN to what you suggested. That did not resolve the issue.

![image](https://github.com/rubygems/rubygems/assets/907178/3ed3de4c-2449-426c-8a61-898b06a17ebc)

However, using `v.tr("-", ".")` resolved the issue. Also, all tests passed with this change here (https://github.com/harshalbhakta/rubygems/pull/1) so I don't think it will break anything else.

![image](https://github.com/rubygems/rubygems/assets/907178/33bea916-1f29-40cd-9b91-4eaaf8373ef2)

Should definitely add this to a method with documentation. Just wanted to be sure that this is the correct way of solving this issue.

Reply:
> If PATTERN is not the best place to do this, just please wrap this into own method with documentation. This way it is just replacing `-` with `.` inline in constructor with no explanation.

Moved it to `normalize_version` method